File: src/main/java/com/lemonzuo/license/jetbrains/generator/server/ServerCodeCertificateGenerator.java
package com.lemonzuo.license.jetbrains.generator.server;

import cn.hutool.core.date.DateField;
import cn.hutool.core.date.DateTime;
import cn.hutool.core.date.DateUtil;
import cn.hutool.crypto.PemUtil;
import com.lemonzuo.license.jetbrains.constant.CertConstant;
import com.lemonzuo.license.jetbrains.util.X509CertUtils;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.bouncycastle.asn1.x500.X500Name;
import org.bouncycastle.asn1.x509.*;
import org.bouncycastle.cert.X509v3CertificateBuilder;
import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
import org.bouncycastle.cert.jcajce.JcaX509ExtensionUtils;
import org.bouncycastle.cert.jcajce.JcaX509v3CertificateBuilder;
import org.bouncycastle.operator.ContentSigner;
import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;

import java.io.FileInputStream;
import java.math.BigInteger;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.cert.X509Certificate;
import java.util.Date;

/**
 * 生成激活码证书
 *
 * @author LemonZuo
 * @create 2024-05-17 00:12
 * step: 2.生成激活码证书
 */
@Slf4j
public class ServerCodeCertificateGenerator {

    @SneakyThrows
    public static void main(String[] args) {
        generate();
    }

    public static void generate() throws Exception {
        // 读取code证书密钥对
        PrivateKey privateKey2048 = PemUtil.readPemPrivateKey(new FileInputStream(CertConstant.PRIVATE_KEY_2048_PATH));
        PublicKey publicKey2048 = PemUtil.readPemPublicKey(new FileInputStream(CertConstant.PUBLIC_KEY_2048_PATH));
        // 读取根CA密钥对
        PrivateKey privateKey4096 = PemUtil.readPemPrivateKey(new FileInputStream(CertConstant.PRIVATE_KEY_4096_PATH));
        PublicKey publicKey4096 = PemUtil.readPemPublicKey(new FileInputStream(CertConstant.PUBLIC_KEY_4096_PATH));

        // 第二种方式
        byte[] codeCertBytes = PemUtil.readPem(new FileInputStream(CertConstant.JETBRAINS_CODE_CA_PATH));
        X509Certificate templateCert = X509CertUtils.createCertificate(codeCertBytes);

        // 1、生成JetProfile CA证书
        standardGenerateCaCert(templateCert, publicKey4096, privateKey4096, CertConstant.CODE_CA_PATH);

        log.info("========== 签发子证书 ==========");
        // 2、签发子证书
        byte[] myCodeCertBytes = PemUtil.readPem(new FileInputStream(CertConstant.CODE_CA_PATH));
        X509Certificate certificate = X509CertUtils.createCertificate(myCodeCertBytes);
        issueChildCertificate(certificate, templateCert, publicKey2048, privateKey4096, CertConstant.CODE_CERT_PATH);
    }

    /**
     * 标准生成JetProfile CA证书
     *
     * @param templateCert JB CA模板证书
     * @param publicKey    4096位 公钥
     * @param privateKey   4096位 私钥
     * @param certPath     证书路径
     * @throws Exception 异常
     */
    public static void standardGenerateCaCert(X509Certificate templateCert, PublicKey publicKey, PrivateKey privateKey, String certPath) throws Exception {
        X500Name issuerName = new X500Name(templateCert.getIssuerX500Principal().getName());
        X500Name subjectName = new X500Name(templateCert.getSubjectX500Principal().getName());
        BigInteger serialNumber = BigInteger.valueOf(System.currentTimeMillis());
        Date notBefore = templateCert.getNotBefore();
        Date notAfter = templateCert.getNotAfter();

        // 创建 SubjectPublicKeyInfo 对象
        SubjectPublicKeyInfo subPubKeyInfo = SubjectPublicKeyInfo.getInstance(publicKey.getEncoded());

        // 使用 JcaX509v3CertificateBuilder 构建 X509v3CertificateBuilder 对象
        // 颁发者信息、序列号、证书生效日期、证书失效日期、证书的主题信息、证书的公钥信息
        JcaX509v3CertificateBuilder certBuilder = new JcaX509v3CertificateBuilder(
                issuerName,
                serialNumber,
                notBefore,
                notAfter,
                subjectName,
                subPubKeyInfo
        );
        // 复制模板证书中的拓展信息到新证书中
        X509CertUtils.copyExtensions(templateCert, certBuilder);

        // 使用 JcaContentSignerBuilder 构建 ContentSigner 对象
        ContentSigner signer = new JcaContentSignerBuilder("SHA256withRSA").build(privateKey);
        X509Certificate cert = new JcaX509CertificateConverter().setProvider("BC").getCertificate(certBuilder.build(signer));
        // 将证书写入 PEM 文件
        X509CertUtils.writeCertificate(cert, certPath);
    }

    /**
     * 签发子证书
     *
     * @param rootCaCert          我的CA根证书
     * @param jetbrainsCaCert JB的CA证书
     * @param publicKey           2048位 公钥
     * @param issuerPrivateKey    4096位 签发者私钥
     * @param certPath            证书路径
     * @throws Exception 异常
     */
    public static void issueChildCertificate(X509Certificate rootCaCert, X509Certificate jetbrainsCaCert, PublicKey publicKey, PrivateKey issuerPrivateKey, String certPath) throws Exception {
        // 根证书 签发者
        X500Name issuerName = new X500Name(rootCaCert.getSubjectX500Principal().getName());
        X500Name subjectName = new X500Name("CN=lemon-from-20220801");
        BigInteger serialNumber = BigInteger.valueOf(System.currentTimeMillis());
        DateTime today = DateUtil.beginOfDay(DateUtil.date());
        // Yesterday
        Date notBefore = DateUtil.offset(today, DateField.DAY_OF_MONTH, -1);
        // 30 years later
        Date notAfter = DateUtil.offset(today, DateField.YEAR, 30);

        // 创建 SubjectPublicKeyInfo 对象
        SubjectPublicKeyInfo subPubKeyInfo = SubjectPublicKeyInfo.getInstance(publicKey.getEncoded());

        // 使用 JcaX509v3CertificateBuilder 构建 X509v3CertificateBuilder 对象
        X509v3CertificateBuilder certBuilder = new JcaX509v3CertificateBuilder(issuerName, serialNumber, notBefore, notAfter, subjectName, subPubKeyInfo);

        // 添加扩展
        JcaX509ExtensionUtils extUtils = new JcaX509ExtensionUtils();
        // entity cert
        certBuilder.addExtension(Extension.basicConstraints, true, new BasicConstraints(false));
        certBuilder.addExtension(Extension.keyUsage, true, new KeyUsage(KeyUsage.digitalSignature | KeyUsage.keyEncipherment));
        certBuilder.addExtension(Extension.extendedKeyUsage, false, new ExtendedKeyUsage(new KeyPurposeId[]{KeyPurposeId.id_kp_serverAuth}));

        SubjectPublicKeyInfo issuerPublicKeyInfo = SubjectPublicKeyInfo.getInstance(jetbrainsCaCert.getPublicKey().getEncoded());
        AuthorityKeyIdentifier authorityKeyIdentifier = extUtils.createAuthorityKeyIdentifier(issuerPublicKeyInfo);
        // 授权密钥标识
        certBuilder.addExtension(Extension.authorityKeyIdentifier, false, authorityKeyIdentifier);

        SubjectPublicKeyInfo subjectPublicKeyInfo = SubjectPublicKeyInfo.getInstance(rootCaCert.getPublicKey().getEncoded());
        SubjectKeyIdentifier subjectKeyIdentifier = extUtils.createSubjectKeyIdentifier(subjectPublicKeyInfo);
        certBuilder.addExtension(Extension.subjectKeyIdentifier, false, subjectKeyIdentifier);

        ContentSigner signer = new JcaContentSignerBuilder("SHA256withRSA").build(issuerPrivateKey);
        X509Certificate cert = new JcaX509CertificateConverter().setProvider("BC").getCertificate(certBuilder.build(signer));
        // 将证书写入 PEM 文件
        X509CertUtils.writeCertificate(cert, certPath);
    }

}

File: jetbrains/server/generator/code/ServerCodeCertificateGenerator.go
package code

import (
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"crypto/x509/pkix"
	"encoding/pem"
	"license/jetbrains/server"
	"license/logger"
	"log"
	"math/big"
	"os"
	"time"
)

type ServerCodeCertificateGenerator struct {
}

func (s *ServerCodeCertificateGenerator) generate() {
	// privateKey2048 := readPrivateKey(server.PRIVATE_KEY_2048_PATH)
	publicKey2048 := s.readPublicKey(server.PUBLIC_KEY_2048_PATH)
	privateKey4096 := s.readPrivateKey(server.PRIVATE_KEY_4096_PATH)
	publicKey4096 := s.readPublicKey(server.PUBLIC_KEY_4096_PATH)

	templateCert := s.readCertificate(server.JETBRAINS_CODE_CA_PATH)

	s.standardGenerateCaCert(templateCert, publicKey4096, privateKey4096, server.CODE_CA_PATH)

	log.Println("========== 签发子证书 ==========")
	certificate := s.readCertificate(server.CODE_CA_PATH)
	s.issueChildCertificate(certificate, templateCert, publicKey2048, privateKey4096, server.CODE_CERT_PATH)
}

func (s *ServerCodeCertificateGenerator) readPrivateKey(path string) *rsa.PrivateKey {
	keyPEM, err := os.ReadFile(path)
	if err != nil {
		logger.Error("Failed to read private key", err)
		return nil
	}
	block, _ := pem.Decode(keyPEM)
	privateKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
	if err != nil {
		logger.Error("Failed to parse private key", err)
		return nil
	}
	return privateKey
}

func (s *ServerCodeCertificateGenerator) readPublicKey(path string) *rsa.PublicKey {
	keyPEM, err := os.ReadFile(path)
	if err != nil {
		logger.Error("Failed to read public key", err)
		return nil
	}
	block, _ := pem.Decode(keyPEM)
	publicKey, err := x509.ParsePKCS1PublicKey(block.Bytes)
	if err != nil {
		logger.Error("Failed to parse public key", err)
		return nil
	}
	return publicKey
}

func (s *ServerCodeCertificateGenerator) readCertificate(path string) *x509.Certificate {
	certPEM, err := os.ReadFile(path)
	if err != nil {
		logger.Error("Failed to read certificate", err)
		return nil
	}
	block, _ := pem.Decode(certPEM)
	cert, err := x509.ParseCertificate(block.Bytes)
	if err != nil {
		logger.Error("Failed to parse certificate", err)
	}
	return cert
}

func (s *ServerCodeCertificateGenerator) standardGenerateCaCert(templateCert *x509.Certificate, publicKey *rsa.PublicKey, privateKey *rsa.PrivateKey, certPath string) {
	serialNumber := big.NewInt(time.Now().UnixNano())
	notBefore := templateCert.NotBefore
	notAfter := templateCert.NotAfter

	certTemplate := x509.Certificate{
		SerialNumber: serialNumber,
		Subject:      templateCert.Subject,
		Issuer:       templateCert.Issuer,
		NotBefore:    notBefore,
		NotAfter:     notAfter,
		KeyUsage:     x509.KeyUsageCertSign | x509.KeyUsageCRLSign,
		IsCA:         true,
	}

	certDER, err := x509.CreateCertificate(rand.Reader, &certTemplate, templateCert, publicKey, privateKey)
	if err != nil {
		logger.Error("Failed to create certificate", err)
		return
	}
	s.writeCertificate(certDER, certPath)
}

func (s *ServerCodeCertificateGenerator) issueChildCertificate(rootCaCert, jetbrainsCaCert *x509.Certificate, publicKey *rsa.PublicKey, issuerPrivateKey *rsa.PrivateKey, certPath string) {
	// issuerName := rootCaCert.Subject
	subjectName := pkix.Name{
		CommonName: "CN=lemon-from-20220801",
	}

	serialNumber := big.NewInt(time.Now().UnixNano())
	notBefore := time.Now().AddDate(0, 0, -1)
	notAfter := time.Now().AddDate(30, 0, 0)

	certTemplate := x509.Certificate{
		SerialNumber:          serialNumber,
		Subject:               subjectName,
		NotBefore:             notBefore,
		NotAfter:              notAfter,
		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
		BasicConstraintsValid: true,
		IsCA:                  false,
	}

	authorityKeyId, err := x509.MarshalPKIXPublicKey(jetbrainsCaCert.PublicKey)
	if err != nil {
		logger.Error("Failed to marshal public key", err)
		return
	}
	subjectKeyId, err := x509.MarshalPKIXPublicKey(rootCaCert.PublicKey)
	if err != nil {
		logger.Error("Failed to marshal public key", err)
		return
	}
	certTemplate.SubjectKeyId = subjectKeyId
	certTemplate.AuthorityKeyId = authorityKeyId

	certDER, err := x509.CreateCertificate(rand.Reader, &certTemplate, rootCaCert, publicKey, issuerPrivateKey)
	if err != nil {
		logger.Error("Failed to create certificate", err)
		return
	}
	s.writeCertificate(certDER, certPath)
}

func (s *ServerCodeCertificateGenerator) writeCertificate(certDER []byte, path string) {
	certPEM := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: certDER})
	if err := os.WriteFile(path, certPEM, 0644); err != nil {
		logger.Error("Failed to write certificate", err)
	}
}
